File: apps/desktop/src/lib/insights/conversation.ts
// Persistent conversation DTOs (issue #102/#110, ADR 0031) — the frontend
// mirror of `crates/capture-types/src/conversation.rs`. ONE shared conversation
// store backs both doors (Quick Recall and Chat); these types are the wire shape
// of its Tauri commands (`list_conversations` / `get_conversation` /
// `search_conversations` / `save_conversation_turn` / `delete_conversation`).
//
// `toolActivities` / `sources` are opaque JSON the frontend round-trips, so they
// are typed `unknown` here and re-narrowed where consumed.

/** A lightweight conversation row for the history list. */
export interface ConversationSummary {
  conversationId: string;
  title: string;
  /** The door that created it: `"quick_recall"` | `"chat"`. */
  origin: string;
  createdAtMs: number;
  updatedAtMs: number;
  turnCount: number;
  /** The first turn's question, truncated. */
  preview: string;
}

/** One persisted question/answer turn, in `turnIndex` order. */
export interface ConversationTurn {
  turnIndex: number;
  question: string;
  answer: string;
  /** The model's reasoning ("thinking") text for this turn, or null when none. */
  reasoning: string | null;
  /** Render-ready parsed answer blocks (the backend-owned render model); null
   *  for a legacy turn the backend parses from `answer` on read. The
   *  get_conversation command always populates this, so the UI never parses. */
  blocks: AnswerBlock[] | null;
  /** Opaque per-turn tool-activity log (JSON array round-tripped by the UI). */
  toolActivities: unknown;
  /** Opaque per-turn Answer Sources (JSON array round-tripped by the UI). */
  sources: unknown;
  /** `"streaming"` | `"done"` | `"error"`. */
  phase: string;
  errorMessage: string | null;
  seededResultCount: number | null;
  createdAtMs: number;
  updatedAtMs: number;
}

/** A fully-hydrated persisted conversation: metadata + every turn. */
export interface Conversation {
  conversationId: string;
  title: string;
  origin: string;
  createdAtMs: number;
  updatedAtMs: number;
  /** Pinned engine provider id, or null/absent when unpinned (global default). */
  provider?: string | null;
  /** Pinned model id within `provider`, or null/absent when unpinned. */
  model?: string | null;
  turns: ConversationTurn[];
}

/** The conversation door that owns a saved turn: the Insights Chat workspace or
 *  the Quick Recall launcher (both now persist to the shared store — #111). */
export type ConversationOrigin = "chat" | "quick_recall";

/** The payload sent to `save_conversation_turn` (upserts the row + the turn). */
export interface SaveConversationTurnRequest {
  conversationId: string;
  title: string;
  origin: ConversationOrigin;
  turnIndex: number;
  question: string;
  answer: string;
  toolActivities: unknown;
  sources: unknown;
  phase: string;
  errorMessage: string | null;
  seededResultCount: number | null;
}

// ── Render-ready chat view model (issue #110, Slice 1) ───────────────────────
// The BACKEND-OWNED render model for a streaming Ask AI turn, mirroring the Rust
// types in `crates/capture-types/src/conversation.rs`. The backend decides what
// a turn looks like (phases, answer blocks, reasoning, tool activity); the
// frontend ONLY renders. There is no codegen — these must agree field-for-field
// with the Rust side, which a serde round-trip test guards.
//
// Option fields that Rust SKIPS when absent (item/block options) are typed
// optional here (`?`); the TurnView's nullable fields always serialize (as null)
// and are typed `T | null`.

/** One row of a horizontal-bar answer block. */
export interface BarsItem {
  label: string;
  value: number;
  sublabel?: string;
}

/** One claim/finding in a dossier block. `confidence` is a 0..1 score. */
export interface DossierItem {
  subject?: string | null;
  statement: string;
  confidence: number;
}

/** One interval in a timeline block. `end` absent = open/instant interval. */
export interface TimelineItem {
  label: string;
  start: string;
  end?: string | null;
  app?: string | null;
  category?: string | null;
}

/** One render-ready answer block, discriminated on `kind`.
 *
 * `prose` carries RAW markdown — the markdown→HTML pass stays on the frontend
 * (AnswerProse). The graphical variants carry already-parsed data, so the UI no
 * longer re-parses fenced JSON. */
export type AnswerBlock =
  | { kind: "prose"; markdown: string }
  | { kind: "bars"; title?: string | null; items: BarsItem[] }
  | { kind: "dossier"; items: DossierItem[] }
  | { kind: "timeline"; title?: string | null; items: TimelineItem[] };

/** One recorded brokered tool call rendered in the turn's activity rail.
 *  `kind` mirrors the Rust `String` (a frontend `AskToolKind` union member). */
export interface ToolActivityEntry {
  kind: string;
  label: string;
  /** The app the call was scoped to (bundle id or display name), if any. */
  app?: string | null;
  /** A resolved icon path for `app`, when the backend found one. */
  appIconPath?: string | null;
}

/** The full render-ready view of ONE Ask AI turn. The backend owns every field;
 *  the frontend only renders. `phase` is the lifecycle string. `sources` is the
 *  opaque Answer-Sources JSON the UI round-trips. The nullable fields always
 *  serialize (as null), hence `T | null` rather than optional. */
export interface TurnView {
  turnIndex: number;
  question: string;
  /** `"seeding" | "thinking" | "streaming" | "done" | "error"`. */
  phase: string;
  blocks: AnswerBlock[];
  reasoning: string | null;
  toolActivities: ToolActivityEntry[];
  liveActivity: ToolActivityEntry | null;
  sources: unknown;
  errorMessage: string | null;
  seededResultCount: number | null;
}

/** A versioned snapshot of a turn's `TurnView`. `version` lets the frontend
 *  discard stale snapshots that race with live `TurnUpdate`s. */
export interface TurnSnapshot {
  conversationId: string;
  version: number;
  view: TurnView;
}

/** One incremental mutation to a `TurnView`, discriminated on `op`. The backend
 *  streams these as a turn progresses; the frontend applies them to its local
 *  view. (`liveActivity` with `entry: null` CLEARS the live line.) */
export type TurnUpdate =
  | { op: "phase"; phase: string }
  | { op: "appendProse"; text: string }
  | { op: "openBlock"; block: AnswerBlock }
  | { op: "reasoning"; text: string }
  | { op: "toolActivity"; entry: ToolActivityEntry }
  | { op: "liveActivity"; entry: ToolActivityEntry | null }
  | { op: "sources"; sources: unknown }
  | { op: "error"; message: string }
  | { op: "done" };

/** The `ask_ai_update` event payload (emitted by Slice 4): a versioned, indexed
 *  `TurnUpdate` for a conversation. Emitted ad-hoc as JSON on the Rust side, so
 *  it has no capture-types struct — this TS type is the sole wire contract. */
export interface AskAiUpdateEvent {
  conversationId: string;
  version: number;
  turnIndex: number;
  update: TurnUpdate;
}

// ── Ask AI availability + Answer Sources ─────────────────────────────────────
// The legacy per-token `ask_ai_*` event shapes (status/delta/reasoning/done/
// error/source) are gone: BOTH doors (Chat and Quick Recall) now consume the
// single versioned `ask_ai_update` transport (AskAiUpdateEvent above), and the
// backend no longer emits the legacy events. Availability + the cited Answer
// Source shape remain, shared across doors.

export interface AskAiAvailability {
  available: boolean;
  reason: string | null;
}

/** One cited Answer Source (a captured frame or audio segment). */
export interface AskAiSource {
  kind: "frame" | "audio";
  frameId: number | null;
  audioSegmentId: number | null;
  appName: string | null;
  windowTitle: string | null;
  startedAt: string;
  endedAt: string;
  sourceKind: "microphone" | "system" | null;
  spanStartMs?: number | null;
  alignedFrameId?: number | null;
}

// ── Per-thread engine pin (set_conversation_engine) ──────────────────────────
// A conversation can be pinned to one engine identity `{provider, model}` from
// the provider-tagged model pool (ADR 0034). The pin is identified by a connected
// provider **instance id** (`AiProviderConfig.id` — equal to the kind id for the
// first instance of a kind, suffixed for additional same-kind instances) plus the
// bare rig-core model id, matching the Rust single resolver
// (`resolve_engine_config`): thread pin → feature override → global default.

import type {
  AiProviderConfig,
  AiRuntimeModel,
  AiRuntimeSettings,
} from "$lib/types/recording";

/** One selectable engine for the per-thread picker. */
export interface PinnableEngine {
  /** Stable engine-pin provider string passed to `set_conversation_engine`. */
  provider: string;
  /** Bare rig-core model id passed to `set_conversation_engine`. */
  model: string;
  /** Human label, e.g. "Anthropic · claude-haiku-4-5". */
  label: string;
}

/** A short, legible label for a model id: the tail after the last "/",
 *  so long routing paths (accounts/.../deepseek-v4-pro) read cleanly. The
 *  full id stays available for a title= tooltip. */
export function shortModelLabel(model: string): string {
  const trimmed = model.trim();
  const tail = trimmed.split("/").pop()?.trim() ?? "";
  return tail.length > 0 ? tail : trimmed;
}

/** A short, friendly label for one provider KIND id. Returns the raw string for
 *  an unknown id (e.g. a `kind-N` instance id with no provider context). */
export function engineProviderLabel(provider: string): string {
  switch (provider) {
    case "anthropic":
      return "Anthropic";
    case "openai":
      return "OpenAI";
    case "openai_compatible":
      return "OpenAI-compatible";
    case "ollama":
      return "Ollama";
    case "llamafile":
      return "Llamafile";
    default:
      return provider;
  }
}

/** Host component of a base URL, for the auto-derived instance label. */
function baseUrlHost(baseUrl: string): string {
  const trimmed = baseUrl.trim();
  if (!trimmed) return "";
  try {
    return new URL(trimmed).host || trimmed;
  } catch {
    return trimmed;
  }
}

/** Display name for a provider instance (mirrors the Settings card): its user
 *  label, else kind + host, else kind + instance number. Keeps two same-kind
 *  instances distinct even with no label typed, including first-party cloud
 *  (anthropic/openai) which has no base-URL host to disambiguate on. */
export function providerInstanceLabel(provider: AiProviderConfig): string {
  const label = provider.label?.trim() ?? "";
  if (label) return label;
  const kindLabel = engineProviderLabel(provider.kind);
  const host = baseUrlHost(provider.baseUrl ?? "");
  if (host) return `${kindLabel} · ${host}`;
  const suffix = provider.id.startsWith(`${provider.kind}-`)
    ? provider.id.slice(provider.kind.length + 1)
    : "";
  return suffix ? `${kindLabel} (${suffix})` : kindLabel;
}

/** Resolve a provider INSTANCE id to its display label against the connected
 *  provider list. Falls back to the kind label for an unknown/removed instance
 *  (e.g. a pin to a provider that has since been disconnected). */
export function providerLabelById(
  providers: readonly AiProviderConfig[] | null | undefined,
  id: string,
): string {
  const provider = providers?.find((p) => p.id === id);
  return provider ? providerInstanceLabel(provider) : engineProviderLabel(id);
}

/** The global default model's provider id, or null when no default is chosen. */
export function defaultEnginePinProvider(
  settings: AiRuntimeSettings,
): string | null {
  return settings.defaultModel?.provider ?? null;
}

/**
 * The model id the global default resolves to, for LABEL purposes only
 * (e.g. the composer picker's "Default · claude-haiku-4-5"); the backend owns
 * actual resolution. Null when no default model is chosen.
 */
export function defaultEngineModel(settings: AiRuntimeSettings): string | null {
  const model = settings.defaultModel?.model?.trim() ?? "";
  return model.length > 0 ? model : null;
}

/**
 * Map the merged provider-tagged model pool (`ai_runtime_list_models`) into
 * pinnable engines for the per-thread picker, de-duplicated by
 * `(provider, model)`. Free-form model-id entry stays allowed on top of this.
 */
export function pinnableEnginesFromModelPool(
  models: readonly AiRuntimeModel[],
  providers?: readonly AiProviderConfig[] | null,
): PinnableEngine[] {
  const seen = new Set<string>();
  const out: PinnableEngine[] = [];
  for (const entry of models) {
    const model = entry.id.trim();
    if (model.length === 0) continue;
    const key = `${entry.provider} ${model}`;
    if (seen.has(key)) continue;
    seen.add(key);
    out.push({
      provider: entry.provider,
      model,
      label: `${providerLabelById(providers, entry.provider)} · ${model}`,
    });
  }
  return out;
}

File: crates/capture-types/src/conversation.rs
//! Persistent conversation DTOs (issue #102, ADR 0031) — pure serde data shared
//! across the Tauri layer and frontend.
//!
//! ONE shared conversation store backs both doors (Quick Recall and Chat). These
//! mirror the `conversations` / `conversation_turns` tables in the Encrypted
//! Capture Index. They carry no logic; storage + retention/wipe policy live in
//! `crates/app-infra/src/conversation`.
//!
//! Conventions (matching the rest of `capture-types`): structs use
//! `#[serde(rename_all = "camelCase")]`. Timestamps are `i64` unix milliseconds
//! (serialized to camelCase `*AtMs`). `tool_activities` / `sources` are opaque
//! JSON the frontend round-trips, so they are typed `serde_json::Value`.

use serde::{Deserialize, Serialize};

/// One question/answer turn within a [`Conversation`], in `turn_index` order.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ConversationTurn {
    pub turn_index: i64,
    pub question: String,
    pub answer: String,
    /// The model's reasoning ("thinking") text for this turn, or `None` when the
    /// turn carried none (and for legacy turns predating the column).
    pub reasoning: Option<String>,
    /// Render-ready parsed answer blocks for this turn, or `None` for a LEGACY
    /// turn predating the column (whose blocks the backend parses from `answer`
    /// on read). Serializes as `null` when absent (always present on the wire).
    pub blocks: Option<Vec<AnswerBlock>>,
    /// Opaque per-turn tool-activity log the frontend round-trips (JSON array).
    pub tool_activities: serde_json::Value,
    /// Opaque per-turn Answer Sources the frontend round-trips (JSON array).
    pub sources: serde_json::Value,
    /// `'streaming'` | `'done'` | `'error'` (frontend-owned phase string).
    pub phase: String,
    pub error_message: Option<String>,
    pub seeded_result_count: Option<i64>,
    pub created_at_ms: i64,
    pub updated_at_ms: i64,
}

/// A fully-hydrated persisted conversation: its metadata plus every turn in
/// `turn_index` order.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Conversation {
    /// Frontend-generated UUID (the stable cross-restart identity).
    pub conversation_id: String,
    pub title: String,
    /// The door that created it: `'quick_recall'` | `'chat'`.
    pub origin: String,
    pub created_at_ms: i64,
    pub updated_at_ms: i64,
    /// The pinned engine provider id (e.g. `"anthropic"` | `"openai"`), or
    /// `None` when unpinned (use the global default engine).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    /// The pinned model id within [`Self::provider`], or `None` when unpinned.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    pub turns: Vec<ConversationTurn>,
}

/// A lightweight conversation row for the history list: metadata plus a turn
/// count and a short preview (the first turn's question, truncated).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ConversationSummary {
    pub conversation_id: String,
    pub title: String,
    pub origin: String,
    pub created_at_ms: i64,
    pub updated_at_ms: i64,
    pub turn_count: i64,
    pub preview: String,
}

// ── Render-ready chat view model (issue #110, Slice 1) ───────────────────────
// These are the BACKEND-OWNED render model for a streaming Ask AI turn. The
// backend decides what a turn looks like (its phases, answer blocks, reasoning,
// tool activity); the frontend only RENDERS them. This replaces the old scheme
// where the frontend re-parsed raw markdown fences (`mnema-bars` etc.) into chart
// data — that parse now lives in the backend, which emits typed [`AnswerBlock`]s.
//
// These types carry no behaviour. The TS mirror in
// `apps/desktop/src/lib/insights/conversation.ts` must agree field-for-field;
// there is no codegen, so the round-trip tests below pin the exact wire shapes.

/// One row of a horizontal-bar answer block.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct BarsItem {
    pub label: String,
    pub value: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sublabel: Option<String>,
}

/// One claim/finding in a dossier answer block. `confidence` is a 0..=1 score.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DossierItem {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subject: Option<String>,
    pub statement: String,
    pub confidence: f64,
}

/// One interval in a timeline answer block. `start`/`end` are RFC3339-ish
/// strings the frontend formats; `end` absent means an open/instant interval.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct TimelineItem {
    pub label: String,
    pub start: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub end: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
}

/// One render-ready block of a turn's answer, internally tagged on `kind`.
///
/// `Prose` carries RAW markdown — the markdown→HTML pass stays on the frontend
/// (in `AnswerProse`). The graphical variants (`Bars`/`Dossier`/`Timeline`)
/// carry already-parsed data, so the frontend no longer re-parses fenced JSON.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum AnswerBlock {
    /// Raw markdown prose: `{ "kind": "prose", "markdown": "..." }`.
    Prose { markdown: String },
    /// A horizontal-bar chart block.
    Bars {
        #[serde(skip_serializing_if = "Option::is_none")]
        title: Option<String>,
        items: Vec<BarsItem>,
    },
    /// A dossier of claims/findings with confidence scores.
    Dossier { items: Vec<DossierItem> },
    /// A timeline of labelled intervals.
    Timeline {
        #[serde(skip_serializing_if = "Option::is_none")]
        title: Option<String>,
        items: Vec<TimelineItem>,
    },
}

/// One recorded brokered tool call rendered in the turn's activity rail.
///
/// `kind` is a frontend-known union string (`search` | `timeline` | `show_text`
/// | `other`), kept as a `String` to mirror the TS `AskToolKind` without forcing
/// the backend to model it as an enum.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ToolActivityEntry {
    pub kind: String,
    pub label: String,
    /// The app the call was scoped to (bundle id or display name), if any.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app: Option<String>,
    /// A resolved icon path for [`Self::app`], when the backend could find one.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub app_icon_path: Option<String>,
}

/// The full render-ready view of ONE Ask AI turn. The backend owns every field;
/// the frontend only renders. `phase` is the lifecycle string
/// (`"seeding" | "thinking" | "streaming" | "done" | "error"`). `sources` is the
/// same opaque Answer-Sources JSON the frontend round-trips on a persisted turn.
///
/// Unlike the item/block option fields (which are SKIPPED when absent), this
/// view's nullable fields always serialize — as JSON `null` when empty — so the
/// frontend can mirror them as `T | null` rather than optional.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct TurnView {
    pub turn_index: i64,
    pub question: String,
    pub phase: String,
    pub blocks: Vec<AnswerBlock>,
    pub reasoning: Option<String>,
    pub tool_activities: Vec<ToolActivityEntry>,
    pub live_activity: Option<ToolActivityEntry>,
    /// Opaque Answer Sources JSON the frontend round-trips (a JSON array).
    pub sources: serde_json::Value,
    pub error_message: Option<String>,
    pub seeded_result_count: Option<i64>,
}

/// A versioned snapshot of a turn's [`TurnView`] for a given conversation. The
/// `version` lets the frontend ignore stale snapshots that race with live
/// [`TurnUpdate`]s.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct TurnSnapshot {
    pub conversation_id: String,
    pub version: u64,
    pub view: TurnView,
}

/// One incremental mutation to a [`TurnView`], internally tagged on `op`. The
/// backend emits a stream of these as a turn streams; the frontend applies them
/// to its local view. (Slice 4 wires the emit path; these are types only.)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "op", rename_all = "camelCase")]
pub enum TurnUpdate {
    /// Advance the turn's lifecycle phase.
    Phase { phase: String },
    /// Append streamed text to the current open prose block.
    AppendProse { text: String },
    /// Open a new answer block (closing any current prose block).
    OpenBlock { block: AnswerBlock },
    /// Append streamed reasoning ("thinking") text.
    Reasoning { text: String },
    /// Record a completed tool call in the activity rail.
    ToolActivity { entry: ToolActivityEntry },
    /// Set or CLEAR the transient "live" activity line. `entry: null` clears it,
    /// so the `entry` key always serializes (never skipped).
    LiveActivity { entry: Option<ToolActivityEntry> },
    /// Replace the turn's Answer Sources with new opaque JSON.
    Sources { sources: serde_json::Value },
    /// Fail the turn with a message.
    Error { message: String },
    /// Mark the turn complete.
    Done,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn answer_block_prose_exact_shape() {
        let block = AnswerBlock::Prose {
            markdown: "hi".to_string(),
        };
        assert_eq!(
            serde_json::to_value(&block).unwrap(),
            json!({ "kind": "prose", "markdown": "hi" })
        );
    }

    #[test]
    fn answer_block_bars_exact_shape() {
        let block = AnswerBlock::Bars {
            title: Some("Top apps".to_string()),
            items: vec![BarsItem {
                label: "Editor".to_string(),
                value: 42.0,
                sublabel: Some("2h".to_string()),
            }],
        };
        assert_eq!(
            serde_json::to_value(&block).unwrap(),
            json!({
                "kind": "bars",
                "title": "Top apps",
                "items": [
                    { "label": "Editor", "value": 42.0, "sublabel": "2h" }
                ]
            })
        );
    }

    #[test]
    fn answer_block_bars_skips_absent_options() {
        // No title; item without a sublabel — both keys should be absent.
        let block = AnswerBlock::Bars {
            title: None,
            items: vec![BarsItem {
                label: "Editor".to_string(),
                value: 1.5,
                sublabel: None,
            }],
        };
        assert_eq!(
            serde_json::to_value(&block).unwrap(),
            json!({
                "kind": "bars",
                "items": [ { "label": "Editor", "value": 1.5 } ]
            })
        );
    }

    #[test]
    fn answer_block_dossier_and_timeline_shapes() {
        let dossier = AnswerBlock::Dossier {
            items: vec![DossierItem {
                subject: None,
                statement: "ships fast".to_string(),
                confidence: 0.5,
            }],
        };
        assert_eq!(
            serde_json::to_value(&dossier).unwrap(),
            json!({
                "kind": "dossier",
                "items": [ { "statement": "ships fast", "confidence": 0.5 } ]
            })
        );

        let timeline = AnswerBlock::Timeline {
            title: None,
            items: vec![TimelineItem {
                label: "Coding".to_string(),
                start: "2026-06-13T10:00:00Z".to_string(),
                end: None,
                app: None,
                category: None,
            }],
        };
        assert_eq!(
            serde_json::to_value(&timeline).unwrap(),
            json!({
                "kind": "timeline",
                "items": [
                    { "label": "Coding", "start": "2026-06-13T10:00:00Z" }
                ]
            })
        );
    }

    #[test]
    fn turn_update_exact_shapes() {
        assert_eq!(
            serde_json::to_value(TurnUpdate::AppendProse {
                text: "abc".to_string()
            })
            .unwrap(),
            json!({ "op": "appendProse", "text": "abc" })
        );

        assert_eq!(
            serde_json::to_value(TurnUpdate::Done).unwrap(),
            json!({ "op": "done" })
        );

        // None must serialize the `entry` key as null (clear the live line).
        assert_eq!(
            serde_json::to_value(TurnUpdate::LiveActivity { entry: None }).unwrap(),
            json!({ "op": "liveActivity", "entry": null })
        );

        assert_eq!(
            serde_json::to_value(TurnUpdate::Phase {
                phase: "streaming".to_string()
            })
            .unwrap(),
            json!({ "op": "phase", "phase": "streaming" })
        );
    }

    #[test]
    fn turn_view_serializes_camel_case_keys() {
        let view = TurnView {
            turn_index: 3,
            question: "what did I do?".to_string(),
            phase: "done".to_string(),
            blocks: vec![AnswerBlock::Prose {
                markdown: "stuff".to_string(),
            }],
            reasoning: None,
            tool_activities: vec![ToolActivityEntry {
                kind: "search".to_string(),
                label: "Searched".to_string(),
                app: None,
                app_icon_path: None,
            }],
            live_activity: None,
            sources: json!([]),
            error_message: None,
            seeded_result_count: None,
        };
        let value = serde_json::to_value(&view).unwrap();
        let obj = value.as_object().unwrap();
        // Nullable view fields serialize (as null), not skipped.
        assert!(obj.contains_key("turnIndex"));
        assert!(obj.contains_key("toolActivities"));
        assert!(obj.contains_key("liveActivity"));
        assert!(obj.contains_key("errorMessage"));
        assert!(obj.contains_key("seededResultCount"));
        assert_eq!(obj["liveActivity"], json!(null));
        assert_eq!(obj["errorMessage"], json!(null));
        assert_eq!(obj["seededResultCount"], json!(null));
        // The tool-activity entry omits its absent app/appIconPath options.
        assert_eq!(
            obj["toolActivities"],
            json!([ { "kind": "search", "label": "Searched" } ])
        );
    }

    #[test]
    fn turn_snapshot_round_trips_and_shape() {
        let snapshot = TurnSnapshot {
            conversation_id: "conv-1".to_string(),
            version: 7,
            view: TurnView {
                turn_index: 0,
                question: "q".to_string(),
                phase: "thinking".to_string(),
                blocks: vec![],
                reasoning: Some("hmm".to_string()),
                tool_activities: vec![],
                live_activity: Some(ToolActivityEntry {
                    kind: "timeline".to_string(),
                    label: "Building timeline".to_string(),
                    app: Some("com.example.app".to_string()),
                    app_icon_path: Some("/tmp/icon.png".to_string()),
                }),
                sources: json!([{ "kind": "frame" }]),
                error_message: Some("boom".to_string()),
                seeded_result_count: Some(5),
            },
        };

        let value = serde_json::to_value(&snapshot).unwrap();
        let obj = value.as_object().unwrap();
        assert!(obj.contains_key("conversationId"));
        assert!(obj.contains_key("version"));
        assert!(obj.contains_key("view"));

        let round_tripped: TurnSnapshot = serde_json::from_value(value).unwrap();
        assert_eq!(round_tripped, snapshot);
    }
}
